odl/discr/discr_mappings.py

```diff
@@ -467,219 +482,307 @@ def _call(self, x, out=None):
-
         for i, p in enumerate(coord_vecs):
             if not np.asarray(p).ndim == 1:
                 raise ValueError('The points in dimension {} must be '
                                  '1-dimensional'.format(i))
-            if not values.shape[i] == len(p):
+            if values.shape[i] != len(p):
                 raise ValueError('There are {} points and {} values in '
                                  'dimension {}'.format(len(p),
                                                        values.shape[i], i))
 
-        # TODO: use the following code to determine upstream if a different
-        # interpolator needs to be used
-
-#        for i, p in enumerate(points):
-#            if not np.all(np.diff(p) > 0.):
-#                raise ValueError("The points in dimension {} must be strictly"
-#                                 " ascending".format(i))
         self.grid = tuple([np.asarray(p) for p in coord_vecs])
         self.values = values
+        self.type = typ
 
     def __call__(self, x, out=None):
         """Do the interpolation.
 
-        Modified for in-place evaluation support and without method
-        choice. Evaluation points are to be given as an array with
-        shape (n, dim), where n is the number of points.
+        Parameters
+        ----------
+        x : meshgrid or `numpy.ndarray`
+            Evaluation points of the interpolator
+        out : `numpy.ndarray`, optional
+            Array to which the results are written. Needs to have
+            correct shape according to input ``x``.
+
+        Returns
+        -------
+        out : `numpy.ndarray`
+            Interpolated values. If ``out`` was given, the returned
+            object is a reference to it.
         """
-        ndim = len(self.grid)
-        if x.ndim != 2:
-            raise ValueError('x has {} axes instead of 2.'.format(x.ndim))
-
-        if x.shape[0] != ndim:
-            raise ValueError('x has axis 1 with length {} instead '
-                             'of the grid dimension {}.'
-                             ''.format(x.shape[0], ndim))
+        if self.type == 'array':
+            # Make a (1, n) array from one with shape (n,)
+            ndim = len(self.grid)
+            if ndim == 1 and x.ndim == 1:
+                x = x[None, :]
+            if x.ndim != 2:
+                raise ValueError('x has {} axes instead of 2.'.format(x.ndim))
+            if x.shape[0] != ndim:
+                raise ValueError('x has axis 1 with length {} instead '
+                                 'of the grid dimension {}.'
+                                 ''.format(x.shape[0], ndim))
+            out_shape = (x.shape[1],)
+        else:
+            if len(x) != len(self.grid):
+                raise ValueError('number of vectors in x is {} instead of '
+                                 'the grid dimension {}.'
+                                 ''.format(len(x), len(self.grid)))
+            if len(x) == 1:
```

Comment:
This is done all the time, how about we make a utility for it? (and make an issue on numpy that `broadcast` should work with a single argument.)